compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/LIRGenerator.java

```diff
@@ -449,29 +451,55 @@ public Variable emitForeignCall(ForeignCallLinkage linkage, LIRFrameState frameS
 
     @Override
     public void emitStrategySwitch(JavaConstant[] keyConstants, double[] keyProbabilities, LabelRef[] keyTargets, LabelRef defaultTarget, Variable value) {
+
         int keyCount = keyConstants.length;
         SwitchStrategy strategy = SwitchStrategy.getBestStrategy(keyProbabilities, keyConstants, keyTargets);
+
+        double minDensity = 1 / Math.sqrt(strategy.getAverageEffort());
+
+        Optional<Hasher> hasher = hasherFor(keyConstants, minDensity);
+
+        double hashTableSwitchDensity = hasher.map(h -> keyCount / (double) h.cardinality()).orElse(0d);
+
         long valueRange = keyConstants[keyCount - 1].asLong() - keyConstants[0].asLong() + 1;
         double tableSwitchDensity = keyCount / (double) valueRange;
+
         /*
          * This heuristic tries to find a compromise between the effort for the best switch strategy
```

Comment:
Do we still have a policy problem here?  The existing policy is hard to make sense of but it seems like it's not really making comparable trade offs between table switch and the other forms.  The table forms are only billed for their space requirements but the other forms are billed for their dynamic work and not for their code size.  Binary search for instance has a log scale effort but doubles in size each time which doesn't seem quite fair.

Reply:
yes, this heuristic needs to be improved. I'm planning to create a new PR with a better one.